File: bdg-eng-ist-ondc-qa-main/bdg-eng-ist-ondc-qa-main/ondc-signature-multi-language/python/test_signature.py
"""
Test Suite for Python ONDC Signature Implementation
"""

import json
import os
import sys
import requests
from ondc_signature import ONDCSignature


def load_test_data():
    """Load test credentials and payloads"""
    with open('/shared/test-credentials.json', 'r') as f:
        creds = json.load(f)
    with open('/shared/test-payloads.json', 'r') as f:
        payloads = json.load(f)
    return creds, payloads


def test_key_generation(creds):
    """Test 1: Key Generation"""
    print("\n" + "="*80)
    print("TEST 1: KEY GENERATION")
    print("="*80)
    
    uat = creds['uat']
    private_key_seed = bytes.fromhex(uat['private_key_seed_hex'])
    
    signer = ONDCSignature(
        uat['subscriber_id'],
        uat['unique_key_id'],
        private_key_seed
    )
    
    print(f"✅ Private key seed: {uat['private_key_seed_hex'][:32]}...")
    print(f"✅ Public key (base64): {signer.public_key_b64[:40]}...")
    
    # Save for cross-language comparison
    result = {
        "language": "Python",
        "public_key": signer.public_key_b64,
        "test": "key_generation",
        "status": "PASS"
    }
    
    with open('/reports/python-keys.json', 'w') as f:
        json.dump(result, f, indent=2)
    
    return signer


def test_digest_generation(signer, payloads):
    """Test 2: Digest Generation"""
    print("\n" + "="*80)
    print("TEST 2: DIGEST GENERATION")
    print("="*80)
    
    results = []
    for test_case in payloads['test_cases']:
        digest = signer._create_digest(test_case['body'])
        print(f"✅ {test_case['name']}: {digest[:40]}...")
        results.append({
            "test_name": test_case['name'],
            "digest": digest
        })
    
    # Save for cross-language comparison
    with open('/reports/python-digests.json', 'w') as f:
        json.dump({"language": "Python", "digests": results}, f, indent=2)


def test_signature_generation(signer, creds, payloads):
    """Test 3: Signature Generation with Fixed Timestamps"""
    print("\n" + "="*80)
    print("TEST 3: SIGNATURE GENERATION (Fixed Timestamps)")
    print("="*80)
    
    fixed_ts = creds['test_fixed_timestamp']
    created = fixed_ts['created']
    expires = fixed_ts['expires']
    
    results = []
    for test_case in payloads['test_cases']:
        auth_header, digest_header, _, _ = signer.generate_signature_header(
            body=test_case['body'],
            created=created,
            expires=expires
        )
        
        print(f"\n✅ {test_case['name']}:")
        print(f"   Auth: {auth_header[:80]}...")
        print(f"   Digest: {digest_header[:80]}...")
        
        results.append({
            "test_name": test_case['name'],
            "authorization": auth_header,
            "digest": digest_header
        })
    
    # Save for cross-language comparison
    with open('/reports/python-signatures.json', 'w') as f:
        json.dump({"language": "Python", "signatures": results}, f, indent=2)


def test_live_api(signer, creds):
    """Test 4: Live API Call to UAT - v3.0/lookup"""
    print("\n" + "="*80)
    print("TEST 4: LIVE API CALL TO UAT - v3.0/lookup")
    print("="*80)
    
    uat = creds['uat']
    endpoint = f"{uat['base_url']}{uat['lookup_endpoint']}"
    
    # Lookup payload - proper format for v3.0/lookup
    test_payload = {
        "subscriber_id": uat['subscriber_id'],
        "domain": "ONDC:RET10"
    }
    
    auth_header, digest_header, _, _ = signer.generate_signature_header(body=test_payload)
    
    # ⚠️ CRITICAL: Must use pre-serialized JSON matching digest calculation
    # Using json=test_payload causes different serialization than digest calculation
    request_body_str = json.dumps(test_payload, separators=(', ', ': '), sort_keys=True, ensure_ascii=False)
    
    headers = {
        'Content-Type': 'application/json',
        'Authorization': auth_header,
        'Digest': digest_header
    }
    
    print(f"📡 Endpoint: {endpoint}")
    print(f"📦 Payload: {request_body_str}")
    print(f"🔑 Authorization: {auth_header}")
    print(f"🔐 Digest: {digest_header}")
    
    try:
        response = requests.post(
            endpoint,
            data=request_body_str,  # Use data= with pre-serialized string, not json=
            headers=headers,
            timeout=10
        )
        
        print(f"\n✅ Response Status: {response.status_code}")
        
        if response.status_code in [200, 201, 202]:
            resp_data = response.json()
            if isinstance(resp_data, list) and len(resp_data) > 0:
                print("🎉 SUCCESS: Signature verified! Participant found in registry!")
                print(f"   Participant data retrieved: {json.dumps(resp_data[0], indent=2)[:200]}...")
                result = {"status": "PASS", "code": response.status_code, "message": "Participant found (200)"}
            elif isinstance(resp_data, dict):
                error_code = resp_data.get('error', {}).get('code', '')
                if error_code == '1001':
                    print("✅ SUCCESS: Signature verified, but participant not found in domain (1001)")
                    result = {"status": "PASS", "code": response.status_code, "message": "Auth OK, domain mismatch"}
                else:
                    print(f"✅ SUCCESS: Signature verified! (error: {error_code})")
                    result = {"status": "PASS", "code": response.status_code}
            else:
                print("✅ SUCCESS: Signature accepted by UAT API!")
                result = {"status": "PASS", "code": response.status_code}
        elif response.status_code == 404:
            print("⚠️  Participant not found (404) - but signature was accepted")
            print("   This proves authentication works; participant just needs to be registered")
            result = {"status": "PASS", "code": response.status_code, "message": "Auth OK, participant not found"}
        elif response.status_code == 401:
            print("❌ FAIL: 401 Unauthorized - Signature verification failed")
            result = {"status": "FAIL", "code": 401, "message": "Signature verification failed"}
        else:
            print(f"⚠️  Unexpected status: {response.status_code}")
            result = {"status": "WARN", "code": response.status_code}
            
        print(f"Response: {response.text}")
        
    except Exception as e:
        print(f"❌ API call failed: {e}")
        result = {"status": "ERROR", "message": str(e)}
    
    # Save result
    with open('/reports/python-api-test.json', 'w') as f:
        json.dump({"language": "Python", "result": result}, f, indent=2)


def main():
    """Run all tests"""
    print("╔" + "="*78 + "╗")
    print("║" + " PYTHON ONDC SIGNATURE IMPLEMENTATION TEST SUITE ".center(78) + "║")
    print("╚" + "="*78 + "╝")
    
    try:
        # Load test data
        creds, payloads = load_test_data()
        
        # Run tests
        signer = test_key_generation(creds)
        test_digest_generation(signer, payloads)
        test_signature_generation(signer, creds, payloads)
        
        # Ensure participant is registered before API test
        print("\n📋 Ensuring participant is registered before API test...")
        try:
            import subprocess
            subprocess.run(['python', '/shared/ensure_participant_registered.py'], check=False)
        except Exception as reg_error:
            print(f"⚠️  Registration check failed: {reg_error}")
            print(f"💡 Continuing anyway - authentication will still be tested")
        
        test_live_api(signer, creds)
        
        print("\n" + "="*80)
        print("✅ ALL PYTHON TESTS COMPLETED")
        print("="*80 + "\n")
        
        return 0
        
    except Exception as e:
        print(f"\n❌ TEST SUITE FAILED: {e}")
        import traceback
        traceback.print_exc()
        return 1


if __name__ == "__main__":
    sys.exit(main())

File: bdg-eng-ist-ondc-qa-main/bdg-eng-ist-ondc-qa-main/ondc-signature-multi-language/php/test_signature.php
<?php
/**
 * Test Suite for PHP ONDC Signature Implementation
 */

require_once 'ONDCSignature.php';

function loadTestData() {
    $creds = json_decode(file_get_contents('/shared/test-credentials.json'), true);
    $payloads = json_decode(file_get_contents('/shared/test-payloads.json'), true);
    return [$creds, $payloads];
}

function testKeyGeneration($creds) {
    echo "\n" . str_repeat('=', 80) . "\n";
    echo "TEST 1: KEY GENERATION\n";
    echo str_repeat('=', 80) . "\n";
    
    $uat = $creds['uat'];
    $privateKeySeedHex = $uat['private_key_seed_hex'];
    
    $signer = new ONDCSignature(
        $uat['subscriber_id'],
        $uat['unique_key_id'],
        $privateKeySeedHex
    );
    
    echo "✅ Private key seed: " . substr($privateKeySeedHex, 0, 32) . "...\n";
    echo "✅ Public key (base64): " . substr($signer->getPublicKey(), 0, 40) . "...\n";
    
    // Save for cross-language comparison
    $result = [
        'language' => 'PHP',
        'public_key' => $signer->getPublicKey(),
        'test' => 'key_generation',
        'status' => 'PASS'
    ];
    
    file_put_contents('/reports/php-keys.json', json_encode($result, JSON_PRETTY_PRINT));
    
    return $signer;
}

function testDigestGeneration($signer, $payloads) {
    echo "\n" . str_repeat('=', 80) . "\n";
    echo "TEST 2: DIGEST GENERATION\n";
    echo str_repeat('=', 80) . "\n";
    
    $results = [];
    foreach ($payloads['test_cases'] as $testCase) {
        $reflection = new ReflectionClass($signer);
        $method = $reflection->getMethod('createDigest');
        $method->setAccessible(true);
        $digest = $method->invoke($signer, $testCase['body']);
        
        echo "✅ {$testCase['name']}: " . substr($digest, 0, 40) . "...\n";
        
        $results[] = [
            'test_name' => $testCase['name'],
            'digest' => $digest
        ];
    }
    
    file_put_contents('/reports/php-digests.json', json_encode([
        'language' => 'PHP',
        'digests' => $results
    ], JSON_PRETTY_PRINT));
}

function testSignatureGeneration($signer, $creds, $payloads) {
    echo "\n" . str_repeat('=', 80) . "\n";
    echo "TEST 3: SIGNATURE GENERATION (Fixed Timestamps)\n";
    echo str_repeat('=', 80) . "\n";
    
    $fixedTs = $creds['test_fixed_timestamp'];
    $created = $fixedTs['created'];
    $expires = $fixedTs['expires'];
    
    $results = [];
    foreach ($payloads['test_cases'] as $testCase) {
        list($authHeader, $digestHeader) = $signer->generateSignatureHeader(
            $testCase['body'], $created, $expires
        );
        
        echo "\n✅ {$testCase['name']}:\n";
        echo "   Auth: " . substr($authHeader, 0, 80) . "...\n";
        echo "   Digest: " . substr($digestHeader, 0, 80) . "...\n";
        
        $results[] = [
            'test_name' => $testCase['name'],
            'authorization' => $authHeader,
            'digest' => $digestHeader
        ];
    }
    
    file_put_contents('/reports/php-signatures.json', json_encode([
        'language' => 'PHP',
        'signatures' => $results
    ], JSON_PRETTY_PRINT));
}

function testLiveAPI($signer, $creds) {
    echo "\n" . str_repeat('=', 80) . "\n";
    echo "TEST 4: LIVE API CALL TO UAT - v3.0/lookup\n";
    echo str_repeat('=', 80) . "\n";
    
    $uat = $creds['uat'];
    $endpoint = $uat['base_url'] . $uat['lookup_endpoint'];
    
    // Lookup payload - proper format for v3.0/lookup
    $testPayload = [
        'subscriber_id' => $uat['subscriber_id'],
        'domain' => 'ONDC:RET10'
    ];
    
    list($authHeader, $digestHeader) = $signer->generateSignatureHeader($testPayload);
    
    // ⚠️ CRITICAL: Must use pre-serialized JSON matching digest calculation
    $requestBodyStr = $signer->serializeBody($testPayload);
    
    echo "📡 Endpoint: {$endpoint}\n";
    echo "📦 Payload: {$requestBodyStr}\n";
    echo "🔑 Authorization: {$authHeader}\n";
    echo "🔐 Digest: {$digestHeader}\n";
    
    // Create HTTP context with headers
    $context = stream_context_create([
        'http' => [
            'method' => 'POST',
            'header' => "Content-Type: application/json\r\n" .
                       "Authorization: {$authHeader}\r\n" .
                       "Digest: {$digestHeader}\r\n",
            'content' => $requestBodyStr,
            'timeout' => 10,
            'ignore_errors' => true
        ],
        'ssl' => [
            'verify_peer' => false,
            'verify_peer_name' => false
        ]
    ]);
    
    $result = [];
    try {
        $response = file_get_contents($endpoint, false, $context);
        $statusLine = $http_response_header[0] ?? '';
        preg_match('/\d{3}/', $statusLine, $matches);
        $statusCode = (int)($matches[0] ?? 0);
        
        echo "\n✅ Response Status: {$statusCode}\n";
        
        if ($statusCode >= 200 && $statusCode < 300) {
            if (strpos(trim($response), '[') === 0) {
                echo "✅ SUCCESS: Signature verified! Participant found in registry!\n";
                $result = ['status' => 'PASS', 'code' => $statusCode, 'participant_found' => true];
            } elseif (strpos($response, '"1001"') !== false) {
                echo "✅ SUCCESS: Signature verified, but participant not found in domain (1001)\n";
                $result = ['status' => 'PASS', 'code' => $statusCode, 'message' => 'Auth OK, domain mismatch'];
            } else {
                echo "✅ SUCCESS: Signature accepted by UAT API!\n";
                $result = ['status' => 'PASS', 'code' => $statusCode];
            }
            echo "Response: {$response}\n";
        } elseif ($statusCode === 404) {
            echo "✅ SUCCESS: Signature accepted (404 means participant not found, but auth worked)\n";
            $result = ['status' => 'PASS', 'code' => 404, 'message' => 'Auth OK, participant not found'];
        } elseif ($statusCode === 401) {
            echo "❌ FAIL: 401 Unauthorized - Signature verification failed\n";
            $result = ['status' => 'FAIL', 'code' => 401, 'message' => 'Signature verification failed'];
        } else {
            echo "⚠️  Unexpected status: {$statusCode}\n";
            $result = ['status' => 'WARN', 'code' => $statusCode];
        }
        
        echo "Response: {$response}\n";
        
    } catch (Exception $e) {
        echo "❌ API call failed: " . $e->getMessage() . "\n";
        $result = ['status' => 'ERROR', 'message' => $e->getMessage()];
    }
    
    file_put_contents('/reports/php-api-test.json', json_encode([
        'language' => 'PHP',
        'result' => $result
    ], JSON_PRETTY_PRINT));
}

function main() {
    echo "╔" . str_repeat('=', 78) . "╗\n";
    echo "║" . str_pad(' PHP ONDC SIGNATURE IMPLEMENTATION TEST SUITE ', 78, ' ', STR_PAD_BOTH) . "║\n";
    echo "╚" . str_repeat('=', 78) . "╝\n";
    
    try {
        list($creds, $payloads) = loadTestData();
        
        $signer = testKeyGeneration($creds);
        testDigestGeneration($signer, $payloads);
        testSignatureGeneration($signer, $creds, $payloads);
        testLiveAPI($signer, $creds);
        
        echo "\n" . str_repeat('=', 80) . "\n";
        echo "✅ ALL PHP TESTS COMPLETED\n";
        echo str_repeat('=', 80) . "\n\n";
        
        exit(0);
        
    } catch (Exception $e) {
        echo "\n❌ TEST SUITE FAILED: " . $e->getMessage() . "\n";
        echo $e->getTraceAsString() . "\n";
        exit(1);
    }
}

main();
